File: src/cpu/o3/add_pred/base_add_pred.cc
#ifndef __CPU_PRED_ADD_PRED_BASE_ADD_PRED_CC__
#define __CPU_PRED_ADD_PRED_BASE_ADD_PRED_CC__

#include "cpu/o3/add_pred/base_add_pred.hh"

BaseAddPred::BaseAddPred(const BaseAddPredParams &params)
    : SimObject(params),
      confidenceSaturation(params.confidence_saturation),
      confidenceThreshold(params.confidence_threshold),
      confidenceUpStep(params.confidence_up_step),
      confidenceDownStep(params.confidence_down_step)
{

}

BaseAddPred::~BaseAddPred()
{

}

Addr
BaseAddPred::predictFromPC(Addr pc, int runAhead)
{
    return 0;
}

void
BaseAddPred::updatePredictor(Addr realAddr, Addr pc,
                             InstSeqNum seqNum,
                             int packetSize)
{

}

int
BaseAddPred::getPacketSize(Addr pc)
{
    return 0;
}

Stats::Group*
BaseAddPred::getStatGroup()
{
    return nullptr;
}

#endif //__CPU_PRED_ADD_PRED_BASE_ADD_PRED_CC__

File: src/cpu/o3/add_pred/BaseAddPred.py
from m5.defines import buildEnv
from m5.params import *
from m5.proxy import *

from m5.objects.BaseCPU import BaseCPU
from m5.objects.FUPool import *
from m5.objects.O3Checker import O3Checker
from m5.objects.BranchPredictor import *
from m5.SimObject import SimObject

class BaseAddPred(SimObject):
    type = 'BaseAddPred'
    cxx_header='cpu/o3/add_pred/base_add_pred.hh'

    confidence_saturation = Param.Int(10,
                            "Maximum confidence achievable for prediction")
    confidence_threshold = Param.Int(8,
                            "Threshold to begin prediction at")
    confidence_up_step = Param.Int(1,
                            "Gain of confidence on correct prediction")
    confidence_down_step = Param.Int(4,
                            "Loss of confidence on wrong prediction")

class SimplePred(BaseAddPred):
    type = 'SimplePred'
    cxx_header='cpu/o3/add_pred/simple_pred.hh'

class DeltaPred(BaseAddPred):
    type = 'DeltaPred'
    cxx_header='cpu/o3/add_pred/delta_pred.hh'